cli: Teach `persist-nic-names` to pin via kargs too

jlebon opened the pull request:
Pinning by kargs has the additional benefit of also covering the initramfs, which is relevant if networking is necessary there (for e.g. a Tang-pinned rootfs), and network kargs are in use that rely on a specific interface name.

To keep it simpler, we don't have nmstate itself call out to rpm-ostree. Instead, a new `--kargs-out` flag is added where nmstate can write out the kargs that the MCO should add/remove.

Related: https://github.com/openshift/machine-config-operator/pull/3684
Related: https://github.com/openshift/machine-config-operator/pull/3706

jlebon committed:
cli: Teach `persist-nic-names` to pin via kargs too

Pinning by kargs has the additional benefit of also covering the
initramfs, which is relevant if networking is necessary there (for
e.g. a Tang-pinned rootfs), and network kargs are in use that rely on a
specific interface name.

To keep it simpler, we don't have nmstate itself call out to rpm-ostree.
Instead, a new `--kargs-out` flag is added where nmstate can write out
the kargs that the MCO should add/remove.

Related: https://github.com/openshift/machine-config-operator/pull/3684
Related: https://github.com/openshift/machine-config-operator/pull/3706

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon commented:
I think before merging this, we should have the MCO code ready to make sure we're all on the same page wrt the approach.

cgwalters commented on rust/src/cli/ncl.rs:
```diff
@@ -343,6 +343,15 @@ fn main() {
+                        .long("kargs-out")
+                        .takes_value(true)
+                        .help(
+                            "When pinning, write kargs to append; \
```
Maybe worth saying explicitly "kargs to append, separated by a single space"?

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -110,8 +112,10 @@ pub(crate) fn run_persist_immediately(
             changed |=
-                persist_iface_name_via_systemd_link(root, mac, iface.name())?;
+                persist_iface_name_via_systemd_link(root, mac, iface_name)?;
+            kargs.push(format!("ifname={iface_name}:{mac}"));
```
Since we have this in two places, maybe worth a `fn format_ifname_karg(iface_name: &str, mac: &str) -> String` ?

jlebon commented:
Updated!

cgwalters reviewed (approved):
This looks sane to me!

cathay4t commented on rust/src/cli/persist_nic.rs:
```diff
@@ -110,8 +112,10 @@ pub(crate) fn run_persist_immediately(
             changed |=
-                persist_iface_name_via_systemd_link(root, mac, iface.name())?;
+                persist_iface_name_via_systemd_link(root, mac, iface_name)?;
+            kargs.push(format_ifname_karg(iface_name, mac));
```
Can we print out the kernel arguments in `dry-run` also?

jlebon replied:
I did that and a few other tweaks as well. :)

jlebon committed:
cli: Mention kargs in persist_nic doc and minor tweaks

Add kernel arguments to the `persist_nic` module docstring. Fix a typo
in the predictable ifname function name. Don't capitalize systemd at the
beginning of sentences.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon committed:
cli: Simplify and make consistent logging in `persist-nic-names`

In the pinning path, we used "would", which would be more appropriate
for dry run mode, but we also used it in non-dry run mode. In the
cleanup path, we used "will" only in dry run mode.

Since "will" works for both dry run mode and non-dry run mode, tweak
things so we always say "will". This simplifies things since we don't
need different logging for each mode.

In the cleanup path, log when we actually remove the link file for
consistency with the pinning path.

While we're here, merge the kernel arg and link file logging.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon committed:
cli: Strengthen karg checking for net.ifnames=0

A karg like `foonet.ifnames=0` will make this function think that
predictable ifnames are disabled. We need the check to be word boundary-
aware.

Split the kernel cmdline on whitespace and check for each element.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon committed:
cli: Split out `has_any_kargs` helper function

Prep for checking for another karg.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon committed:
cli: Only append `ifname` kargs if `rd.neednet` is used

If networking isn't required in the initramfs, then there's no need to
append `ifname` kargs. Key off of the `rd.neednet` karg to know if that's
the case.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

cathay4t commented:
@jlebon is this PR still in draft state?

jlebon commented:
> @jlebon is this PR still in draft state?

Yes, let's have the MCO side ready before merging this. (One can do a local build/scratch build of nmstate with this PR and test it in the MCO CI as done in https://github.com/openshift/machine-config-operator/pull/3684).

The MCO bits would still need to be adapted in 4.14 first before backporting to 4.13. If we want the nmstate operator to get unblocked more quickly, it'd probably be better to pursue https://github.com/openshift/machine-config-operator/pull/3706#issuecomment-1559672426.

Edit: opened https://github.com/openshift/machine-config-operator/pull/3716

cgwalters commented:
FWIW I think Jonathan is likely to go on leave soon, I'll take over handling this PR.

cgwalters commented:
It's not clear to me what in the MCO would need to change before merging this?

cgwalters commented:
This patch on top of this series re-adds support for `--inspect`, does it look OK?

```From edff40fddd0228eadf1219919763e583f07bbf37 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Tue, 30 May 2023 16:27:42 -0400
Subject: [PATCH] persist_nic: Readd --inspect

---
 rust/src/cli/ncl.rs         |  8 ++++++++
 rust/src/cli/persist_nic.rs | 35 +++++++++++++++++++++++------------
 2 files changed, 31 insertions(+), 12 deletions(-)

diff --git a/rust/src/cli/ncl.rs b/rust/src/cli/ncl.rs
index 7cd59e58..7800ce59 100644
--- a/rust/src/cli/ncl.rs
+++ b/rust/src/cli/ncl.rs
@@ -334,6 +334,12 @@ fn main() {
                         .takes_value(false)
                         .help("Only output changes that would be made"),
                 )
+                .arg(
+                    clap::Arg::new("INSPECT")
+                        .long("inspect")
+                        .takes_value(false)
+                        .help("Output information about prior state, if any"),
+                )
                 .arg(
                     clap::Arg::new("CLEAN_UP")
                         .long("cleanup")
@@ -453,6 +459,8 @@ fn main() {
                 }
             } else if matches.try_contains_id("CLEAN_UP").unwrap_or_default() {
                 persist_nic::PersistAction::CleanUp
+            } else if matches.try_contains_id("INSPECT").unwrap_or_default() {
+                persist_nic::PersistAction::Inspect
             } else {
                 persist_nic::PersistAction::Save
             };
diff --git a/rust/src/cli/persist_nic.rs b/rust/src/cli/persist_nic.rs
index 71df8bd4..b045b476 100644
--- a/rust/src/cli/persist_nic.rs
+++ b/rust/src/cli/persist_nic.rs
@@ -48,6 +48,8 @@ pub(crate) enum PersistAction {
     Save,
     /// Print what we would do in Save mode
     DryRun,
+    /// Summarize previous state
+    Inspect,
     /// Remove link files not required
     CleanUp,
     /// Print what we would do in clean up mode
@@ -69,17 +71,21 @@ pub(crate) fn run_persist_immediately(
 ) -> Result<String, CliError> {
     let dry_run = match action {
         PersistAction::Save => false,
-        PersistAction::DryRun => true,
+        // Inspecting counts as a dry run
+        PersistAction::DryRun | PersistAction::Inspect => true,
         PersistAction::CleanUp => return clean_up(root, kargsfile, false),
         PersistAction::CleanUpDryRun => return clean_up(root, kargsfile, true),
     };
+    let inspect = matches!(action, PersistAction::Inspect);
 
     if is_predictable_ifname_disabled() {
         log::info!(
             "systemd predictable network interface name is disabled \
             by kernel argument `net.ifnames=0`, will do nothing"
         );
-        return Ok("".to_string());
+        if !inspect {
+            return Ok("".to_string());
+        }
     }
 
     let stamp_path = Path::new(root)
@@ -87,7 +93,9 @@ pub(crate) fn run_persist_immediately(
         .join(NMSTATE_PERSIST_STAMP);
     if stamp_path.exists() {
         log::info!("{} exists; nothing to do", stamp_path.display());
-        return Ok("".to_string());
+        if !inspect {
+            return Ok("".to_string());
+        }
     }
 
     let mut kargs: Vec<String> = Vec::new();
@@ -113,6 +121,14 @@ pub(crate) fn run_persist_immediately(
             Some(m) => m,
             None => continue,
         };
+        let file_path = gen_link_file_path(root, iface.name());
+        if file_path.exists() {
+            log::info!(
+                "Network link file {} already exists",
+                file_path.display()
+            );
+            continue;
+        }
         let iface_name = iface.name();
         let karg = format_ifname_karg(iface_name, mac);
         log::info!("Will persist the interface {iface_name} with MAC {mac}");
@@ -120,13 +136,13 @@ pub(crate) fn run_persist_immediately(
             log::info!("Will append kernel argument {karg}");
         }
         if !dry_run {
-            changed |=
-                persist_iface_name_via_systemd_link(root, mac, iface_name)?;
+            persist_iface_name_via_systemd_link(root, mac, iface_name)?;
             if with_kargs {
                 log::info!("Kernel argument {karg} appended");
                 kargs.push(karg);
             }
         }
+        changed = true;
     }
 
     if !changed {
@@ -350,15 +366,10 @@ fn persist_iface_name_via_systemd_link(
     root: &str,
     mac: &str,
     iface_name: &str,
-) -> Result<bool, CliError> {
+) -> Result<(), CliError> {
     let link_dir = Path::new(root).join(SYSTEMD_NETWORK_LINK_FOLDER);
 
     let file_path = gen_link_file_path(root, iface_name);
-    if file_path.exists() {
-        log::info!("Network link file {} already exists", file_path.display());
-        return Ok(false);
-    }
-
     if !link_dir.exists() {
         std::fs::create_dir(&link_dir)?;
     }
@@ -376,7 +387,7 @@ fn persist_iface_name_via_systemd_link(
         "systemd network link file created at {}",
         file_path.display()
     );
-    Ok(true)
+    Ok(())
 }
 
 fn is_nmstate_generated_systemd_link_file(file_path: &PathBuf) -> bool {
-- 
2.40.1

```

cgwalters commented:
Since I can't push to this one, I've opened https://github.com/nmstate/nmstate/pull/2361

cathay4t commented:
Since https://github.com/nmstate/nmstate/pull/2361 is merged. Should we close this one?

cgwalters commented:
Yep!
